src/BuiltInTools/dotnet-watch/HotReload/CompilationHandler.cs

```diff
@@ -0,0 +1,239 @@
+                    // diagnostics mean the latter. In this case, treat it as if we successfully applied the delta.
+
+                    // Calling Workspace.TryApply on an MSBuildWorkspace causes the workspace to write source text changes to disk.
+                    // We'll instead simply keep track of the updated solution.
+                    _currentSolution = updatedSolution;
+
+                    // Figure out how to recover from errors. Currently it seems unrecoverable.
+                    return false;
+                }
+
+                _reporter.Verbose("Unable to apply update.");
+                return false;
+            }
+
+            _editAndContinue.CommitSolutionUpdate();
+            _editAndContinue.EndEditSession();
+
+            // Calling Workspace.TryApply on an MSBuildWorkspace causes the workspace to write source text changes to disk.
+            // We'll instead simply keep track of the updated solution.
+            _currentSolution = updatedSolution;
+
+            return await _deltaApplier.Apply(context, file.FilePath, updates, cancellationToken);
+        }
+
+        private IReadOnlyList<string> GetDiagnostics(Project project, CancellationToken cancellationToken)
+        {
+            if (!project.TryGetCompilation(out var compilation))
+            {
+                return Array.Empty<string>();
+            }
+
+            var compilationDiagnostics = compilation.GetDiagnostics(cancellationToken);
+            if (compilationDiagnostics.IsDefaultOrEmpty)
+            {
+                return Array.Empty<string>();
+            }
+
+            var diagnostics = new List<string>();
+            foreach (var item in compilationDiagnostics.Where(d => d.Severity == DiagnosticSeverity.Error))
+            {
+                diagnostics.Add(CSharpDiagnosticFormatter.Instance.Format(item));
+            }
+
+            return diagnostics;
+        }
+
+        private async ValueTask<bool> EnsureSolutionInitializedAsync()
+        {
+            if (_currentSolution != null)
+            {
+                return true;
+            }
+
+            if (_initializeTask is null)
+            {
+                return false;
+            }
+
+            try
+            {
+                (_currentSolution, _editAndContinue) = await _initializeTask;
+                return true;
+            }
+            catch (Exception ex)
+            {
+                _reporter.Warn(ex.Message);
+                return false;
+            }
+        }
+
+        private async ValueTask<SourceText> GetSourceTextAsync(string filePath)
+        {
+            // FSW events sometimes appear before the file has been completely written to disk. Provide a small delay before we read the file contents
+            // to ensure we are not contending with partial writes or write locks.
+            await Task.Delay(20);
```

Comment:
Is the goal here to maximize the chance of the first call succeeding?

Reply:
Yup. This is obviously very twitchy, but so far it seems to work. I should remove the `Task.Delay` and see if I can make this work.